vshard/log_ratelimit.lua

```diff
@@ -0,0 +1,202 @@
+--         code = <int>,
+--         name = <string>,
+--     }
+--
+local function ratelimit_can_log(limiter, new_entry)
+    local entry_type, code = ratelimit_get_signature(new_entry)
+    assert(entry_type and code)
+    local map = limiter.map
+    return map[entry_type] == nil or map[entry_type][code] == nil
+end
+
+local function ratelimit_suppress_entry(limiter, entry)
+    local entry_type, code = ratelimit_get_signature(entry)
+    assert(entry_type and code)
+    local existing = limiter.map[entry_type][code]
+    existing.suppressed = existing.suppressed + 1
+end
+
+local function ratelimit_add_entry(limiter, new_entry)
+    assert(consts.LOG_RATELIMIT_INTERVAL > 0)
+    local entry_type, code = ratelimit_get_signature(new_entry)
+    assert(entry_type and code)
+    local new_map_entry = {
+        entry = new_entry,
+        suppressed = 0,
+    }
+    local map = limiter.map
+    local map_type = map[entry_type] or {}
+    map_type[code] = new_map_entry
+    map[entry_type] = map_type
+    local new_heap_entry = {
+        deadline = fiber.clock() + consts.LOG_RATELIMIT_INTERVAL,
+        type = entry_type,
+        code = code,
+    }
+    limiter.heap:push(new_heap_entry)
+end
+
+local function ratelimit_flush(limiter)
+    local heap = limiter.heap
+    if heap:count() == 0 then
+        return
+    end
+    local map = limiter.map
+    local current_ts = fiber.clock()
+    while heap:top() and heap:top().deadline <= current_ts do
+        local top = heap:pop()
+        assert(map[top.type] and map[top.type][top.code])
+        local map_entry = map[top.type][top.code]
+        map[top.type][top.code] = nil
+        if not next(map[top.type]) then
+            map[top.type] = nil
+        end
+        if map_entry.suppressed > 0 then
+            local e = map_entry.entry
+            -- Some errors can have no names (e.g. `SocketError`).
+            log.info("Suppressed %d '%s' messages from '%s'",
+                     map_entry.suppressed,
+                     e.name or string.format('%s.%s', e.type, e.code),
+                     limiter.name)
+        end
+    end
+end
+
+local function ratelimit_log_template(log_lvl)
+    return function (limiter, entry, format, ...)
+        ratelimit_flush(limiter)
+        local level = 'verbose'
+        -- luacheck: ignore 542 (empty if branch)
+        if not ratelimit_check_entry(entry) then
+            -- The entry doesn't satisfy requirements, print in verbose level.
+        elseif ratelimit_can_log(limiter, entry) then
+            -- The entry doesn't exist in the limiter, print in requested level
+            -- and add it to the limiter, so that the consequent entries are
+            -- not printed until it's flushed.
+            ratelimit_add_entry(limiter, entry)
+            level = log_lvl
+        else
+            -- The entry should be suppressed, it's in the limiter already.
+            ratelimit_suppress_entry(limiter, entry)
+        end
```

Comment:
The signature is potentially obtained here up to 3 times per log ("check entry" passes, "can log" doesn't, then "suppress" is called). It might be wiser to extract the signature first. And then pass it everywhere instead of the `entry`. So it is calculated only once.

Reply:
If you do this, ping me again, I will have a look.